plugin/trino-sqlserver/src/main/java/io/trino/plugin/sqlserver/SqlServerClient.java

```diff
@@ -207,6 +218,26 @@
 
     private static final int MAX_SUPPORTED_TEMPORAL_PRECISION = 7;
 
+    private static final PredicatePushdownController SQLSERVER_STRING_COLLATION_AWARE_PUSHDOWN = (session, domain) -> {
+        if (domain.isOnlyNull()) {
+            return FULL_PUSHDOWN.apply(session, domain);
+        }
+
+        // we can push down only for equals, however don't know how to overcome not equals, because it's also ok to pushdown
+        if (domain.isSingleValue()) {
```

Comment:
IMO the check should be reversed, i.e. something like

```
if (domain.isNullableSingleValue()) {
    <full_pushdown>
}

if (!domain.getValues().isDiscreteSet()) {
    // Push down of range predicate for varchar/char types could lead to incorrect results when the remote database sorts values differently
    <disable_pushdown>
}

Domain simplifiedDomain = domain.simplify(getDomainCompactionThreshold(session));
if (!simplifiedDomain.getValues().isDiscreteSet()) {
    <disable_pushdown>
}
```

The benefit being that an earlier short circuit without having to spend time merging valuesets if not needed.

Reply:
Fixed